### PR TITLE
romio: add special-case DAOS detection logic

### PR DESCRIPTION
**Pull request by roblatham00**
## Pull Request Description

DAOS relies on fuse to expose a posix-like interface to users.  we don't want to use fuse or posix, but the usual `stat` approach can only tell us a file is on a fuse file system.  

We'll punch a small hole in our abstractions with this, but I don't know a better way:  if daos enabled, and a file comes back as 'fuse', see if that file is actually daos. 

closes pmodels/mpich#6930

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by roblatham00** [comment]
Finally had a chance to test this on Aurora, using my https://github.com/roblatham00/hintdump program.

plain file path: file on DAOS but no prefix given:

```
hintdump default
key = romio_daos_chunk_size      value = 0
key = romio_daos_obj_class       value = OC_UNKNOWN
key = cb_buffer_size             value = 16777216
key = romio_cb_read              value = automatic
key = romio_cb_write             value = automatic
key = cb_nodes                   value = 1
key = romio_no_indep_rw          value = false
key = romio_cb_pfr               value = disable
key = romio_cb_fr_types          value = aar
key = romio_cb_fr_alignment      value = 1
key = romio_cb_ds_threshold      value = 0
key = romio_cb_alltoall          value = automatic
key = ind_rd_buffer_size         value = 4194304
key = romio_ds_read              value = automatic
key = romio_ds_write             value = disable
key = romio_synchronized_flush   value = disabled
key = romio_visibility_immediate  value = true
key = cb_config_list             value = *:1
key = romio_filesystem_type      value = DAOS: ROMIO driver for DAOS
key = mpi_memory_alloc_kinds     value = mpi,system
key = romio_aggregator_list      value = 0
```
a file with a daos: prefix still works:
```
hintdump with prefix
key = romio_daos_chunk_size      value = 0
key = romio_daos_obj_class       value = OC_UNKNOWN
key = cb_buffer_size             value = 16777216
key = romio_cb_read              value = automatic
key = romio_cb_write             value = automatic
key = cb_nodes                   value = 1
key = romio_no_indep_rw          value = false
key = romio_cb_pfr               value = disable
key = romio_cb_fr_types          value = aar
key = romio_cb_fr_alignment      value = 1
key = romio_cb_ds_threshold      value = 0
key = romio_cb_alltoall          value = automatic
key = ind_rd_buffer_size         value = 4194304
key = romio_ds_read              value = automatic
key = romio_ds_write             value = disable
key = romio_synchronized_flush   value = disabled
key = romio_visibility_immediate  value = true
key = cb_config_list             value = *:1
key = romio_filesystem_type      value = DAOS: ROMIO driver for DAOS
key = mpi_memory_alloc_kinds     value = mpi,system
key = romio_aggregator_list      value = 0
```
and we still operate on other kinds of file systems (lustre) as expected;
```
hintdump somewhere else
key = direct_read                value = false
key = direct_write               value = false
key = romio_lustre_co_ratio      value = 1
key = romio_lustre_coll_threshold  value = 0
key = romio_lustre_ds_in_coll    value = enable
key = cb_buffer_size             value = 16777216
key = romio_cb_read              value = automatic
key = romio_cb_write             value = automatic
key = cb_nodes                   value = 1
key = romio_no_indep_rw          value = false
key = romio_cb_pfr               value = disable
key = romio_cb_fr_types          value = aar
key = romio_cb_fr_alignment      value = 1
key = romio_cb_ds_threshold      value = 0
key = romio_cb_alltoall          value = automatic
key = ind_rd_buffer_size         value = 4194304
key = ind_wr_buffer_size         value = 524288
key = romio_ds_read              value = automatic
key = romio_ds_write             value = automatic
key = romio_synchronized_flush   value = disabled
key = romio_visibility_immediate  value = true
key = cb_config_list             value = *:1
key = romio_filesystem_type      value = LUSTRE:
key = mpi_memory_alloc_kinds     value = mpi,system
key = romio_aggregator_list      value = 0
key = striping_unit              value = 1048576
key = striping_factor            value = 2
key = start_iodevice             value = 65535
```

**Comment by roblatham00** [comment]
@mchaarawi what do you think of this approach?

**Comment by roblatham00** [comment]
test:mpich/ch4/most

File: src/mpi/romio/adio/common/ad_fstype.c [hunk 1]
```diff
@@ -41,6 +41,14 @@
 #include "gpfs.h"
 #endif
 
+#ifdef ROMIO_DAOS
+#include <daos_types.h>
+#include <daos_obj_class.h>
+#include <daos_cont.h>
+#include <daos_prop.h>
```
Comment (mchaarawi) [inline src/mpi/romio/adio/common/ad_fstype.c:48]: i believe those first 4 are unnecessary and only daos_uns.h is sufficient

File: src/mpi/romio/adio/common/ad_fstype.c [hunk 2]
```diff
@@ -370,6 +386,23 @@ static int romio_statfs(const char *filename, int64_t * file_id)
...
+    if (*file_id == FUSE_SUPER_MAGIC) {
+        *file_id = UNKNOWN_SUPER_MAGIC;
+#ifdef ROMIO_DAOS
+        int ret;
+        /* DAOS adds a wrinkle here:  fuse overwrites the backend's stat structure
+         * with the fuse filesystem type.  For DAOS, we have to make a special call
+         * to see if a file on FUSE is actually DAOS */
+        struct duns_attr_t attr = { 0 };;
+
+        ret = duns_resolve_path(filename, &attr);
```
Comment (mchaarawi) [inline src/mpi/romio/adio/common/ad_fstype.c:400]: im not too familiar with the internals here of romio_statfs. so some questions:
1) if several ranks open the file collectively, does this end up being called on every rank?
2) if one specified daos: prefix or sets the env variable `ROMIO_FSTYPE_FORCE="daos:"`, is this still called?

I did some local testing and i came to the conclusion that the answers are:
1) no
2) no
which is great, but just wanted to verify

File: src/mpi/romio/adio/common/ad_fstype.c [hunk 3]
```diff
@@ -370,6 +386,23 @@ static int romio_statfs(const char *filename, int64_t * file_id)
...
+    /* fuse is a special case: normally we don't care if a file is on fuse
+     * or not, unless that file is actually a DAOS file */
+    if (*file_id == FUSE_SUPER_MAGIC) {
+        *file_id = UNKNOWN_SUPER_MAGIC;
+#ifdef ROMIO_DAOS
+        int ret;
+        /* DAOS adds a wrinkle here:  fuse overwrites the backend's stat structure
+         * with the fuse filesystem type.  For DAOS, we have to make a special call
+         * to see if a file on FUSE is actually DAOS */
+        struct duns_attr_t attr = { 0 };;
```
Comment (mchaarawi) [inline src/mpi/romio/adio/common/ad_fstype.c:398]: extra ;

File: src/mpi/romio/adio/common/ad_fstype.c (hunk 2 above)
Comment (roblatham00) [reply src/mpi/romio/adio/common/ad_fstype.c:400]: yeah, check out `ADIO_FileSysType_fncall_scalable` which we can call unless NFS is enabled 

(NFS?  yeah... imagine a system where a file system lives on rank 0 or the "master" node and the other nodes in the system NFS mount it..  well it was a bug report back in 2007 i am so old...)

**Comment by roblatham00** [comment]
thanks for the careful review.  I pushed a new version with those suggestions

**Review by mchaarawi — APPROVED** [review]

**Review by raffenet — APPROVED** [review]
LGTM

**Pull request merged**